enclave-runtime/src/rpc/author/mod.rs

```diff
@@ -130,16 +127,38 @@ enum TopSubmissionMode {
 	SubmitWatch,
 }
 
-impl<'a, P> Author<'a, P>
+impl<TopPoolGetter> Author<TopPoolGetter>
 where
-	P: TrustedOperationPool + Sync + Send + 'static,
+	TopPoolGetter: GetTopPool + Sync + Send + 'static,
 {
+	/// Execute a specific task or function on the TOP pool
+	///
+	/// Needed to get a clean way around the mutexes in which the pool are wrapped.
+	/// Other approaches require lots of code duplication for the mutex locking and
+	/// unwrapping.
+	fn execute_on_top_pool<F, E, R>(&self, function_to_execute: F, mutex_error_func: E) -> R
+	where
+		F: FnOnce(&TopPoolGetter::TrustedOperationPool) -> R,
+		E: FnOnce(Error) -> R,
```

Comment:
We cannot return a result, because some of our functions return Futures, not results. If it were `Result` (which was my first try), then yes, we wouldn't need the error mapping function.

Reply:
Another random thought, have you considered:
```
fn execute_on_top_pool<F, E, R>(&self, function_to_execute: F, mutex_error_func: E) -> Result<R, TopPoolError>
```

where the outer result basically expresses if the setup to execute `function_to_execute` was successful and `R` is the result of the passed function.